server/src/test/java/org/elasticsearch/node/NodeTests.java

```diff
@@ -675,4 +679,65 @@ public void testPluggablePersistedClusterStateServiceValidation() throws IOExcep
             }
         }
     }
+
+    public void testInitialClusterStateWaitCancelledOnShutdown() throws Exception {
+
+        Consumer<MockNode> startNode = node -> {
+            try {
+                node.start();
+            } catch (NodeValidationException e) {
+                throw new AssertionError(e);
+            }
+        };
+        Settings baseSettings = Settings.builder()
+            .put(ClusterName.CLUSTER_NAME_SETTING.getKey(), InternalTestCluster.clusterName("single-node-cluster", randomLong()))
+            .put(NetworkModule.TRANSPORT_TYPE_KEY, getTestTransportType())
+            .build();
+        Settings masterSettings = Settings.builder()
+            .put(baseSettings)
+            .put(Environment.PATH_HOME_SETTING.getKey(), createTempDir())
+            .put("node.name", "master_node")
+            .build();
+        List<Class<? extends Plugin>> plugins = basePlugins();
+        plugins.add(MockTransportService.TestPlugin.class);
+
+        try (MockNode masterNode = new MockNode(NodeRoles.masterOnlyNode(masterSettings), plugins)) {
+            startNode.accept(masterNode);
+
+            CountDownLatch reachedBlock = new CountDownLatch(1);
+            CountDownLatch testDone = new CountDownLatch(1);
+
+            var transportService = asInstanceOf(MockTransportService.class, masterNode.injector().getInstance(TransportService.class));
+            transportService.addRequestHandlingBehavior("internal:discovery/request_peers", (handler, request, channel, task) -> {
+                logger.info("blocking peer discovery");
+                reachedBlock.countDown();
+                testDone.await();
```

Comment:
Do we need to block here? I think we can just fall through.

Reply:
I think this test already covers the case we want adequately (particularly if we remove `testDone.await()` and make `INITIAL_STATE_TIMEOUT_SETTING` much longer).

We could also ensure we see the `shutdown began while waiting for initial discovery state` log message after (and only after) the call to `joiningNode.prepareForClose()`